consensus/safety-rules/src/safety_rules_2chain.rs

```diff
@@ -18,7 +18,7 @@ impl SafetyRules {
         &mut self,
         timeout: &TwoChainTimeout,
         timeout_cert: Option<&TwoChainTimeoutCertificate>,
-    ) -> Result<Ed25519Signature, Error> {
+    ) -> Result<bls12381::Signature, Error> {
```

Comment:
using generics is too intrusive across all the structs. it's also not very useful for serialization/deserialization since the type is erased but maybe some future warrior can solve it

Reply:
I figured... Sounds like an unfortunate Rust problem.